libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt

```diff
@@ -12,18 +12,21 @@ import java.io.File
  * Implementation of [VoiceApi] allowing you to retrieve voice instructions.
  */
 internal class MapboxVoiceApi(
-    private val speechProvider: MapboxSpeechProvider,
+    private val speechLoader: MapboxSpeechLoader,
     private val speechFileProvider: MapboxSpeechFileProvider
 ) : VoiceApi {
 
+    init {
+        VoiceInstructionsPredownloadHub.register(speechLoader)
+    }
```

Comment:
As I understand you did in this way to make predownloading enabled for all users by default. Our components responsible for speech finds instance of `MapboxNavigation`, listen for route progress, and download voice instructions under the hood. 
This approach has its benefits, and I guess the main benefit is that it works out of the box for all users.

But there are also some downsides:
1. `SpeeachApi` used to have a very straight forward api: user asks it to generate voice instruction - it generates. It may be surprising for users and the SDK developers that there are some work which is going on under the hood.
2. Those hidden access of mapbox navigation instance seems like a cyclic dependency which makes it harder to understand data flow between navigation core and voice generation: some data comes from mapbox navigation via users and some via hidden access of mapbox navigation.

Maybe we can implement predownloading without hidden dependencies so that it's works explicitly for the users and the SDK developers. 
What if speech API provide `predownload` function + the SDK provide a separate component which analyses route progress and triggers callback where user asks speech api to predownload some instructions. In this case it will be clear for user how the predownloading happens. 
We can additionally provide a higher level API, where all the components are connected and works out of the box.

For me this decision  seems like a tradeoff between easy to integrate vs easy to support/troubleshoot. I would picked the latter.

@dzinad , did you consider similar approaches to the suggested one? What was behind your decision to pick the approach which you have picked?

Reply:
Capturing from an internal discussion.
Providing a solution for all users so that they don't have to do anything on their end is the main reason why I chose this approach. Predownloading instructions fixes a bug when voice instructions are being played too late with low connectivity and it makes sense to fix the bug for everyone and not say: "do this and then the bug will be gone".
We've also discussed an alternative approach. We don't provide a solution under the hood, however, we integrate pre-downloading for users that use high-level APIs and for low-level API users we add a `MapboxSpeechApi#predownload` function and a way of getting instructions that should be predownloaded (for example, via a new observer).
I like this approach as well because it fits into the current NavSDK paradigm (Drop-In vs low-level APIs) but that wouldn't fix the bug for all users under the hood. And I would really like that.
It would be nice to hear more opinions on this discussion. @abhishek1508 @tomaszrybakiewicz @kmadsen @Guardiola31337

Reply:
The DropInUi component exists for audio guidance with `MapboxAudioGuidance`. If we introduce a new top level api, it would be something like `MapboxAudioGuidanceOptions` where you could specify details of preloading - I do not think we should add options yet. 

The rest of the pre-downloaded code would be lower level. I like the idea of adding a function like `MapboxSpeechApi#preload`, make it experimental because we may iterate on parameters. Have you decided how to deal with memory/disk constraints? And then the `MapboxSpeechApi#generate` function would be able to continue to work.

It may be worth looking at image caching libraries for inspiration because the implementation details have been iterated on (with strong opinions) for years. For example, maybe there should be a LRU cache but has voice instructions instead of images? Load it until the cache is full or there are no more instructions for the route.  https://github.com/square/picasso/blob/master/picasso/src/main/java/com/squareup/picasso3/PlatformLruCache.kt

Reply:
> The DropInUi component exists for audio guidance with MapboxAudioGuidance. If we introduce a new top level api, it would be something like MapboxAudioGuidanceOptions where you could specify details of preloading - I do not think we should add options yet.

I don't think we should introduce new API, I think that `MapboxAudioGuidance` should support pre-downloading out of the box.

> The rest of the pre-downloaded code would be lower level. I like the idea of adding a function like MapboxSpeechApi#preload, make it experimental because we may iterate on parameters. Have you decided how to deal with memory/disk constraints? And then the MapboxSpeechApi#generate function would be able to continue to work.

For now caching is handled by Core SDK.
But I'm especially worried about not fixing an issue with playing voice instruction too late out of the box if we add new API. The whole idea came from the fact that there is a bug. Any thoughts on that?

> It may be worth looking at image caching libraries for inspiration because the implementation details have been iterated on (with strong opinions) for years. For example, maybe there should be a LRU cache but has voice instructions instead of images? Load it until the cache is full or there are no more instructions for the route. https://github.com/square/picasso/blob/master/picasso/src/main/java/com/squareup/picasso3/PlatformLruCache.kt

There's been a Review Forum (Week 32) discussion on which approach we should choose for implementing pre-downloading. And I thought we'd come up with an overall plan. Maybe I was wrong but it seemed to me that at the moment what was left is implementation and PR review would be mostly to discuss implementation details and not change the approach in general. Of course there are hundreds of ways it can be implemented and everyone might have their own opinion on what is best. If there is a specific reason why you think that previously chosen approach is bad, let's close this PR and start from the beginning with another round of discussion.

Reply:
> But I'm especially worried about not fixing an issue with playing voice instruction too late out of the box if we add new API. The whole idea came from the fact that there is a bug. Any thoughts on that?

It will, but users will have to adopt new API which predownloads voice instructions

Reply:
There are a lot of things said here that are being missed in translation. We're in agreement on many things that are considered not in agreement.

When looking at the implementation, the only thing I actually disagree with is adding a mechanism that hooks into the MapboxNavigation constructor, that is not needed and introduces a very awkward objects (e.g., LegacyMapboxNavigationInstanceHolder)

Reply:
I'm also not seeing logic for managing or defining the limits for the preloaded instructions. 

How do you know a super long route will not take up all of the disk space? Also how do you clear up the free space?

Reply:
> There are a lot of things said here that are being missed in translation. We're in agreement on many things that are considered not in agreement.

Is there any specific sentence of mine you didn't fully understand? Please share because I'm asking for your opinions in order to have a constructive discussion and that will not be possible if we don't understand each other well.

Reply:
Kind of hard to explain, your responses here https://github.com/mapbox/mapbox-navigation-android/pull/6547#discussion_r1054091115 were not in line with the points i was trying to make. So I feel like my statements were not clear. 

[I'm seeing from other comments](https://github.com/mapbox/mapbox-navigation-android/pull/6547#discussion_r1054452244) that we are all in line, so I'm thinking it's all good. I must have missed Vadzim suggestion, because that is also what I was suggesting.

In regards to memory constraints and cleaning up instructions, we can move that discussion to this thread https://github.com/mapbox/mapbox-navigation-android/pull/6547#discussion_r1054300679

Reply:
The second iteration that takes into account the discussion above can be found here: https://github.com/mapbox/mapbox-navigation-android/pull/6771